Sources/TensorFlow/LazyTensor/LazyTensorTrace.swift

```diff
@@ -0,0 +1,140 @@
+import CTensorFlow
+
+/// A class describing the tensor operations that need to be executed
+/// to evaluate a given `LazyTensorOperation`.
+class LazyTensorTrace {
+    // inputs will be "placeholder" nodes.
+    var inputs: [LazyTensorOperation] = []
+    var inputValues: [TFETensorHandle] = []
+    var operations: [LazyTensorOperation] = []
+    var outputs: [LazyTensorOperation] = []
+    var originalOutputs: [LazyTensorOperation] = []
+
+    init(_ lazyOp: LazyTensorOperation) {
+        // TODO: We only pick operations on which `lazyOp` depends on. Note that
+        // there may be other live tensors that could also be materialized at
+        // this time. e.g.,
+        //   x = a + b
+        //   y = x + c
+        // For `x`, only `a + b` is extracted. One optimization is to also include
+        // `y = x + c` into the trace so that we don't have the overhead of creating
+        // another trace when we need to materialize `y`.
+        //
+        let _ = collectLazyOp(lazyOp)
+        lazyOpsCache.removeAll()
+    }
+
+    func signature() -> String {
+        let inputsDesc = inputs.map { input -> String in
+            let dtypeAttr = input.attrs["dtype"]!
+            return "\(input.nameWithID):\(dtypeAttr)"
+        }
+        let inputDesc = inputsDesc.joined(separator: ",")
+        let outputsDesc = outputs.map { output -> String in
+            return "\(output.nameWithID)"
+        }
+        let outputDesc = outputsDesc.joined(separator: ",")
+        return "lazyTrace_\(operations.count)(\(inputDesc)) -> (\(outputDesc))"
+    }
+
+    private var lazyOpsCache: [ObjectIdentifier: LazyTensorOperation] = [:]
+
+    private func updateCacheAndOperations(
+        _ id: ObjectIdentifier, _ node: LazyTensorOperation) {
+        lazyOpsCache[id] = node
+        node.id = "\(operations.count)"
+        operations.append(node)
+    }
+
+    private func newConstTensor(_ conc: TFETensorHandle) -> LazyTensor {
+        let cTensorHandle = conc._cTensorHandle
+        let result = LazyTensorOperation("Const", 1)
+        let dtype = TensorDataType(TFE_TensorHandleDataType(cTensorHandle))
+        let dtypeAttr = LazyTensorOperation.Attribute.TensorDataTypeValue(dtype)
+        result.attrs = [
+            "dtype": dtypeAttr,
+            "value": LazyTensorOperation.Attribute.ConstTensor(conc)]
+        updateCacheAndOperations(ObjectIdentifier(conc), result)
+        return LazyTensor(_lazy: result, index: 0)
+    }
+
+    private func newPlaceholderTensor(_ conc: TFETensorHandle) -> LazyTensor {
```

Comment:
Is this behaving like a factory method? If so, it's better to start the name with "make". Same for `newConstTensor(_:)` above.

Source: https://swift.org/documentation/api-design-guidelines/
> Begin names of factory methods with “make”, e.g. `x.makeIterator()`.

Reply:
Done.